### PR TITLE
add new "no-wrapper" property to apps

### PR DESCRIPTION
**Pull request by mvo5**
A RFC branch to fix https://bugs.launchpad.net/snapcraft/+bug/1706575

This allows to specify that no shell wrapper for apps should be generated. This is useful for e.g. statically linked binaries that don't need this anyway.

LP: 1706575

### CODE REVIEW

File: schema/snapcraft.yaml [hunk 1]
```diff
@@ -190,6 +190,10 @@ properties:
                 - type: string
                   minLength: 1
                 - type: number
+          no-wrapper:
```
Comment (zyga) [inline schema/snapcraft.yaml]: Suggestion: make this "wrappers" enum that has possible values "legacy" and "none" as we may be able to provide some kind of useful wrappers later on.

Comment (sergiusens) [inline schema/snapcraft.yaml]: Agreed, this should be an array of unique elements like `build-attributes` in parts.

File: integration_tests/snaps/assemble/snapcraft.yaml [hunk 2]
```diff
@@ -10,6 +10,8 @@ apps:
     command: binary1
   binary2:
     command: subdir/binary2
+  binary-no-wrapper:
+    command: subdir/binary3
```
Comment (elopio) [inline integration_tests/snaps/assemble/snapcraft.yaml]: This is missing:
`wrapper: none`

File: integration_tests/test_snap.py [hunk 3]
```diff
@@ -73,6 +73,10 @@ def test_snap(self):
             os.path.join(self.prime_dir, 'bin', 'not-wrapped.wrapper'),
             Not(FileExists()))
 
+        self.assertThat(
+            os.path.join(self.prime_dir, 'bin', 'command-binary3.wrapper'),
```
Comment (elopio) [inline integration_tests/test_snap.py]: The name of the file should be: `command-binary-no-wrapper.wrapper`.

**Comment by mvo5** [comment]
Thanks for the feedback and sorry for my slow reply. I addressed the review feedback now. Please let me know if I can help in any other way.

File: schema/snapcraft.yaml [hunk 4]
```diff
@@ -190,6 +190,13 @@ properties:
                 - type: string
                   minLength: 1
                 - type: number
+          wrapper:
```
Comment (zyga) [inline schema/snapcraft.yaml]: Nice :-)

File: snapcraft/internal/meta.py [hunk 5]
```diff
@@ -371,7 +371,11 @@ def _wrap_apps(self, apps):
             if os.path.splitext(f)[1] == '.desktop':
                 os.remove(os.path.join(gui_dir, f))
         for app in apps:
-            self._wrap_app(app, apps[app])
+            wrapper = "legacy"
+            if "wrapper" in apps[app]:
+                wrapper = apps[app]["wrapper"]
```
Comment (zyga) [inline snapcraft/internal/meta.py]: `wrapper = apps[app].get("wrapper", legacy)` is shorter

Not sure if snapcraft uses anything that can derive the default from the schema.

Comment (kalikiana) [inline snapcraft/internal/meta.py]: Yes. `default: legacy` does exactly that, so no need to do it in code.

**Comment by sergiusens** [comment]
Hi, thanks for the changes, I was thinking of something more like:

```
apps:
    no-wrapper-app:
        command: app
        app-constraints: [no-wrapper]
    wrapped-app
        command: app
```

wdyt?

**Comment by sergiusens** [comment]
Now that I think of it, it `no-wrapper` might even be `no-environment` as we already have plans to get rid of wrappers and make use of `env` for `apps` a lot more.

**Comment by mvo5** [comment]
I updated the PR now to follow the suggestion of Gustavo in https://forum.snapcraft.io/t/telling-snapcraft-to-skip-generating-wrapper-scripts/1635/3 - i.e. it is using: `adapter: {"", "none"}` now.

**Comment by mvo5** [comment]
How are things looking? I updated the code based on the feedback that I got from @niemeyer. Please let me know if there is anything else I can do to help this PR along, I need it for https://github.com/snapcore/snapd/pull/3625 so it woudl be great if it could make it to the next snapcraft release.

**Comment by kyrofa** [comment]
Ping on this one.

**Comment by mvo5** [comment]
Thanks, updated and removed the "" from the enum.

**Pull request merged**